File: luajit_programs_and_containers/building_blocks/thread_bridge/fnv1a/fnv1a.c
/**
 * fnv1a.c — FNV-1a 32-bit hash, proper uint32_t arithmetic.
 *
 * Used by Thread bridge field registry and ChainTree dispatch.
 * Matches the Lua implementation in chain_tree_luajit/lua_dsl/lua_support/fnv1a.lua
 * but without integer overflow concerns.
 */

#include <stdint.h>
#include <stddef.h>

#define FNV_OFFSET_32 0x811C9DC5u
#define FNV_PRIME_32  0x01000193u

uint32_t fnv1a_32(const char *data, size_t len) {
    uint32_t hash = FNV_OFFSET_32;
    for (size_t i = 0; i < len; i++) {
        hash ^= (uint8_t)data[i];
        hash *= FNV_PRIME_32;
    }
    return hash;
}

/* String variant (null-terminated) */
uint32_t fnv1a_32_str(const char *str) {
    uint32_t hash = FNV_OFFSET_32;
    while (*str) {
        hash ^= (uint8_t)*str++;
        hash *= FNV_PRIME_32;
    }
    return hash;
}

/* Schema hash: "file.h:record" format, returns signed int32 for JSON */
int32_t fnv1a_schema_hash(const char *file_name, const char *record_name) {
    /* Build key: "{file_name}.h:{record_name}" inline */
    uint32_t hash = FNV_OFFSET_32;

    /* Hash file_name */
    const char *p = file_name;
    while (*p) { hash ^= (uint8_t)*p++; hash *= FNV_PRIME_32; }

    /* Hash ".h:" */
    hash ^= '.'; hash *= FNV_PRIME_32;
    hash ^= 'h'; hash *= FNV_PRIME_32;
    hash ^= ':'; hash *= FNV_PRIME_32;

    /* Hash record_name */
    p = record_name;
    while (*p) { hash ^= (uint8_t)*p++; hash *= FNV_PRIME_32; }

    return (int32_t)hash;
}

File: luajit_programs_and_containers/building_blocks/thread_bridge/lib/fnv1a.lua
--[[
    fnv1a.lua — LuaJIT FFI wrapper for libfnv1a.so

    Proper uint32_t arithmetic via C. No integer overflow concerns.

    Usage:
        local fnv = require("fnv1a")
        local hash = fnv.hash("hello")           -- uint32
        local hash = fnv.hash_bytes(ptr, len)     -- raw bytes
        local sh   = fnv.schema_hash("file", "record")  -- signed int32
]]

local ffi = require("ffi")

ffi.cdef[[
    uint32_t fnv1a_32(const char *data, size_t len);
    uint32_t fnv1a_32_str(const char *str);
    int32_t  fnv1a_schema_hash(const char *file_name, const char *record_name);
]]

-- Load from same directory as this script, or LD_LIBRARY_PATH
local script_dir = debug.getinfo(1, "S").source:match("^@(.*/)")  or "./"
local ok, C = pcall(ffi.load, script_dir .. "../fnv1a/libfnv1a.so")
if not ok then
    C = ffi.load("fnv1a")
end

local M = {}

--- Hash a string, returns unsigned 32-bit integer.
function M.hash(str)
    return tonumber(C.fnv1a_32(str, #str))
end

--- Hash raw bytes (cdata pointer + length).
function M.hash_bytes(data, len)
    return tonumber(C.fnv1a_32(data, len))
end

--- Hash a null-terminated string.
function M.hash_str(str)
    return tonumber(C.fnv1a_32_str(str))
end

--- Schema hash: fnv1a("{file}.h:{record}"), returns signed int32.
function M.schema_hash(file_name, record_name)
    return tonumber(C.fnv1a_schema_hash(file_name, record_name))
end

M._C = C

return M
